## Team Members:
Vasantha Chandrasekaran and Saahil Hiranandani

## Importing libraries

In [1]:
import numpy as np
import ipywidgets
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
import altair as alt
%matplotlib inline

myJekyllSaveDir = '../assets/json/'


## Importing the dataset
### Steam Game Dataset

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/Saahil87/Saahil87.github.io/main/_data/games-features.csv') # Read the dataset

In [3]:
df

QueryID  ResponseID                  QueryName  \
0           10          10             Counter-Strike   
1           20          20      Team Fortress Classic   
2           30          30              Day of Defeat   
3           40          40         Deathmatch Classic   
4           50          50  Half-Life: Opposing Force   
...        ...         ...                        ...   
13352   567660      567660              Baseball Riot   
13353   567860      567860                  Passage 4   
13354   567940      567940                 Piximalism   
13355   568150      568150                 Technoball   
13356   559250      559250           Ghostlight Manor   

                    ResponseName  ReleaseDate  RequiredAge  DemoCount  \
0                 Counter-Strike   Nov 1 2000            0          0   
1          Team Fortress Classic   Apr 1 1999            0          0   
2                  Day of Defeat   May 1 2003            0          0   
3             Deathmatch Classic   Jun 1 2001            0          0   
4      Half-Life: Opposing Force   Nov 1 1999            0          0   
...                          ...          ...          ...        ...   
13352              Baseball Riot  Jan 17 2017            0          0   
13353                  Passage 4  Dec 13 2016            0          0   
13354                 Piximalism  Sep 26 2019            0          0   
13355                 Technoball  Dec 12 2016            0          0   
13356           Ghostlight Manor  Dec 12 2016            0          0   

       DeveloperCount  DLCCount  Metacritic  ...  \
0                   1         0          88  ...   
1                   1         0           0  ...   
2                   1         0          79  ...   
3                   1         0           0  ...   
4                   1         0           0  ...   
...               ...       ...         ...  ...   
13352               1         0           0  ...   
13353               1         0           0  ...   
13354               1         0           0  ...   
13355               1         0           0  ...   
13356               1         0           0  ...   

                          LegalNotice  \
0                                       
1                                       
2                                       
3                                       
4                                       
...                               ...   
13352  Copyright (c) 2016 10tons Ltd.   
13353   2016 copyright by netmin e.K.   
13354                                   
13355                                   
13356                                   

                                                 Reviews  \
0                                                          
1                                                          
2                                                          
3                                                          
4                                                          
...                                                  ...   
13352                                                      
13353                                                      
13354                                                      
13355                                                      
13356  The real draw here is the gameplay Digital Fut...   

                                      SupportedLanguages  \
0      English French German Italian Spanish Simplifi...   
1                  English French German Italian Spanish   
2                  English French German Italian Spanish   
3                  English French German Italian Spanish   
4                           English French German Korean   
...                                                  ...   
13352         English**languages with full audio support   
13353  English* French Italian German* Spanish Dutch*...   
13354                                            English   
13355 

In [4]:
df.info() #Information about the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13357 entries, 0 to 13356
Data columns (total 78 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   QueryID                      13357 non-null  int64  
 1   ResponseID                   13357 non-null  int64  
 2   QueryName                    13356 non-null  object 
 3   ResponseName                 13357 non-null  object 
 4   ReleaseDate                  13357 non-null  object 
 5   RequiredAge                  13357 non-null  int64  
 6   DemoCount                    13357 non-null  int64  
 7   DeveloperCount               13357 non-null  int64  
 8   DLCCount                     13357 non-null  int64  
 9   Metacritic                   13357 non-null  int64  
 10  MovieCount                   13357 non-null  int64  
 11  PackageCount                 13357 non-null  int64  
 12  RecommendationCount          13357 non-null  int64  
 13  PublisherCount  

## Size of the Dataset:
The Steam Game dataset is large with 13357 rows and 78 columns.The file size is 52.6 MB.
## GitHub limits:
As GitHub limits the size of the files allowed in repositories to not be larger than 50 MB, the best way to upload the file would be removing the commit to minimize performance impact. However,the Steam Game dataset is of size 52.6 MB and GitHub would push files of size less than 100 MB to the repository with a warning. 
Source - https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-large-files-on-github

We could push the files of size less than 100 MB to the repository with a warning. Hence,the dataset is not changed from Part-1 and is of size 52.6 MB.

## Data Manipulation:

In [5]:
df.isnull().sum()

QueryID             0
ResponseID          0
QueryName           1
ResponseName        0
ReleaseDate         0
                   ..
PCRecReqsText       0
LinuxMinReqsText    0
LinuxRecReqsText    0
MacMinReqsText      0
MacRecReqsText      0
Length: 78, dtype: int64

In [6]:
# Replacing empty values with Unspecified
df['QueryName'].fillna('Unspecified',inplace=True)
df['SupportEmail'].fillna('Unspecified',inplace=True)
df['SupportURL'].fillna('Unspecified',inplace=True)
df['LegalNotice'].fillna('Unspecified',inplace=True)

In [7]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13357 entries, 0 to 13356
Data columns (total 78 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   QueryID                      13357 non-null  int64  
 1   ResponseID                   13357 non-null  int64  
 2   QueryName                    13357 non-null  object 
 3   ResponseName                 13357 non-null  object 
 4   ReleaseDate                  13357 non-null  object 
 5   RequiredAge                  13357 non-null  int64  
 6   DemoCount                    13357 non-null  int64  
 7   DeveloperCount               13357 non-null  int64  
 8   DLCCount                     13357 non-null  int64  
 9   Metacritic                   13357 non-null  int64  
 10  MovieCount                   13357 non-null  int64  
 11  PackageCount                 13357 non-null  int64  
 12  RecommendationCount          13357 non-null  int64  
 13  PublisherCount  

In [8]:
df.isnull()

QueryID  ResponseID  QueryName  ResponseName  ReleaseDate  RequiredAge  \
0        False       False      False         False        False        False   
1        False       False      False         False        False        False   
2        False       False      False         False        False        False   
3        False       False      False         False        False        False   
4        False       False      False         False        False        False   
...        ...         ...        ...           ...          ...          ...   
13352    False       False      False         False        False        False   
13353    False       False      False         False        False        False   
13354    False       False      False         False        False        False   
13355    False       False      False         False        False        False   
13356    False       False      False         False        False        False   

       DemoCount  DeveloperCount  DLCCount  Metacritic  ...  LegalNotice  \
0          False           False     False       False  ...        False   
1          False           False     False       False  ...        False   
2          False           False     False       False  ...        False   
3          False           False     False       False  ...        False   
4          False           False     False       False  ...        False   
...          ...             ...       ...         ...  ...          ...   
13352      False           False     False       False  ...        False   
13353      False           False     False       False  ...        False   
13354      False           False     False       False  ...        False   
13355      False           False     False       False  ...        False   
13356      False           False     False       False  ...        False   

       Reviews  SupportedLanguages  Website  PCMinReqsText  PCRecReqsText  \
0        False               False    False          False          False   
1        False               False    False          False          False   
2        False               False    False          False          False   
3        False               False    False          False          False   
4        False               False    False          False          False   
...        ...                 ...      ...            ...            ...   
13352    False               False    False          False          False   
13353    False               False    False          False          False   
13354    False               False    False          False          False   
13355    False               False    False          False          False   
13356    False               False    False          False          False   

       LinuxMinReqsText  LinuxRecReqsText  MacMinReqsText  MacRecReqsText  
0                 False             False           False           False  
1                 False             False           False           False  
2                 False             False           False           False  
3                 False             False           False           False  
4                 False             False           False           False  
...                 ...               ...             ...             ...  
13352             False             False           False           False  
13353             False             False           False           False  
13354             False             False           False           False  
13355             False             False           False           False  
13356             False             False           False           False  

[13357 rows x 78 columns]

In [9]:
df.columns

Index(['QueryID', 'ResponseID', 'QueryName', 'ResponseName', 'ReleaseDate',
       'RequiredAge', 'DemoCount', 'DeveloperCount', 'DLCCount', 'Metacritic',
       'MovieCount', 'PackageCount', 'RecommendationCount', 'PublisherCount',
       'ScreenshotCount', 'SteamSpyOwners', 'SteamSpyOwnersVariance',
       'SteamSpyPlayersEstimate', 'SteamSpyPlayersVariance',
       'AchievementCount', 'AchievementHighlightedCount', 'ControllerSupport',
       'IsFree', 'FreeVerAvail', 'PurchaseAvail', 'SubscriptionAvail',
       'PlatformWindows', 'PlatformLinux', 'PlatformMac', 'PCReqsHaveMin',
       'PCReqsHaveRec', 'LinuxReqsHaveMin', 'LinuxReqsHaveRec',
       'MacReqsHaveMin', 'MacReqsHaveRec', 'CategorySinglePlayer',
       'CategoryMultiplayer', 'CategoryCoop', 'CategoryMMO',
       'CategoryInAppPurchase', 'CategoryIncludeSrcSDK',
       'CategoryIncludeLevelEditor', 'CategoryVRSupport', 'GenreIsNonGame',
       'GenreIsIndie', 'GenreIsAction', 'GenreIsAdventure', 'GenreIsCasual',
       'G

### Formatting the ReleaseDate Column
The 'Release Date' column has dates in the form- 'Month(as string) Date Year'. It is converted to from ('Year- Month-Date).

In [10]:
import datetime as dt

In [11]:
df['ReleaseDate']

0         Nov 1 2000
1         Apr 1 1999
2         May 1 2003
3         Jun 1 2001
4         Nov 1 1999
            ...     
13352    Jan 17 2017
13353    Dec 13 2016
13354    Sep 26 2019
13355    Dec 12 2016
13356    Dec 12 2016
Name: ReleaseDate, Length: 13357, dtype: object

In [12]:
df['ReleaseDate'] = pd.to_datetime(df['ReleaseDate'],errors='coerce',format='%b %d %Y')

In [13]:
df['ReleaseDate']

0       2000-11-01
1       1999-04-01
2       2003-05-01
3       2001-06-01
4       1999-11-01
           ...    
13352   2017-01-17
13353   2016-12-13
13354   2019-09-26
13355   2016-12-12
13356   2016-12-12
Name: ReleaseDate, Length: 13357, dtype: datetime64[ns]

## PLOT-1: Driver Plot
### Plot showing the count of free and paid games according to their month of release


In [14]:
chart_1 = alt.Chart.from_dict({
    "data":{"url":"https://raw.githubusercontent.com/Saahil87/Saahil87.github.io/main/_data/games-features.csv"},
  "mark":"bar",
  "height":500,
  "width":500,
  "encoding":{
    "x":{
      "timeUnit":"month",
      "field": "ReleaseDate",
      "type": "ordinal",
      "title": "Release Month"
    },
    "y":{
        "aggregate":"count",
        "type":"quantitative"
    },
    "color": {"field": "IsFree","type": "ordinal"}
          
  }
            
          })

chart_1

alt.Chart(...)

In [15]:
chart_1.properties(width='container').save(myJekyllSaveDir+'project-chart1.vl.json')

### DRIVER PLOT :
Chart_1, shown above is the "Driver" plot.
#### Description:
Chart_1  is a bar chart that shows the count of records of 'free' and 'paid' games which are of the domain 'True' and 'False' of the 'IsFree' column, with respect to the month of release of that game.

The x-axis includes the release months of the game from January to December. There are few values whose release month is not specified,and they correspond to the unnamed value.

The y-axis includes the count of records of type -'IsFree == True' (dark blue) and 'IsFree== False'(light blue).
The count of records of free and paid games can be calculated and can also be compared according to their release months.
#### Aesthetics:
Labels: The x and y labels are properly defined as 'Release Month' and 'Count of Records' respectively.

Color Scheme: The color of the values is chosen to be default- Blue. The dark and light blue colors represent the count of records of free and paid games respectively.

Scale: The scales representing the free(IsFree-TRUE) and paid games (IsFree-FALSE) and its color scheme is represented at the top-right corner of the driver plot.

## PLOT-2: Driven Plot
## Plot showing the maximum achievement count with the release month of the game

In [16]:
chart_2 = alt.Chart.from_dict({
  "data":{"url":"https://raw.githubusercontent.com/Saahil87/Saahil87.github.io/main/_data/games-features.csv"},
  "mark":"line",
    "height":500,
    "width":500,
  "encoding":{
    "x":{"timeUnit":"month","field": "ReleaseDate","type": "ordinal","title": "Release Month"},
    "y":{"field":"AchievementCount","type":"quantitative","axis":{"title":"Maximum Achievement Count"},"aggregate":"max"}
  }}
)


chart_2

alt.Chart(...)

In [17]:
chart_2.properties(width='container').save(myJekyllSaveDir+'project-chart2.vl.json')

### DRIVEN PLOT :
Chart_2, shown above is the "Driven" plot from the Driver plot.
#### Description:
Chart_2  is a line chart that shows the maximum achievement count of games with respect to the month of release of games in the Steam Game Platform.
The x-axis includes the release months of games from January to December. There are few values whose release month is not specified,and they correspond to the unnamed value.

The y-axis includes the maximum achievement count in games.
The maximum achievement count of games in the Steam Game platform can be calculated and can also be compared according to the release month of the games.
#### Aesthetics:
Labels: The x and y labels are properly defined as 'Release Month' and 'Maximum Achievement Count' respectively.

Color Scheme: The color of the line is chosen to be default- Blue. 

Scale: The scales and color scheme is appropriate to the line graph.

## Non-Interactive Dashboard
The dashboard below is non-interactive. It is just to visualize the driver plot and the driven plot with no interactivity. 

In [18]:
chart = alt.HConcatChart(hconcat=[chart_1,chart_2])
chart

alt.HConcatChart(...)

The chart above shows the 'Count of free and paid records vs Release Month' and the 'Maximum Achievement Count vs Release Month' as two charts arranged side by side. The number of free and paid games and the maximum achievement count in each month of the release of the games can be observed from the above chart.

## Dashboard:
The above two charts are linked to create a dashboard. Interactivity is achieved using the 'brush' tool and the changes made due to the selection of a particular interval in chart-1 are reflected in chart-2.

In [22]:
brush = alt.selection_interval(encodings=['x','y'])
chart_1 = alt.Chart.from_dict({
    "data":{"url":"https://raw.githubusercontent.com/Saahil87/Saahil87.github.io/main/_data/games-features.csv"},
  "mark":"bar",
  "height":500,
  "width":500,
  "encoding":{
    "x":{
      "timeUnit":"month",
      "field": "ReleaseDate",
      "type": "ordinal",
      "title": "Release Month"
    },
    "y":{
        "aggregate":"count",
        "type":"quantitative"
    },
    "color": {"field": "IsFree","type": "ordinal"}
          
  }
            
          }).add_selection(
        brush
    )
chart_2 = alt.Chart.from_dict({
  "data":{"url":"https://raw.githubusercontent.com/Saahil87/Saahil87.github.io/main/_data/games-features.csv"},
  "mark":"line",
    "height":500,
    "width":500,
  "encoding":{
    "x":{"timeUnit":"month","field": "ReleaseDate","type": "ordinal","title": "Release Month"},
    "y":{"field":"AchievementCount","type":"quantitative","axis":{"title":" Maximum Achievement Count"},"aggregate":"max"}
  }}
).transform_filter(
        brush
    )
chart = chart_1 | chart_2

In [23]:
chart

alt.HConcatChart(...)

In [24]:
chart.save(myJekyllSaveDir+'project-dashboard.vl.json')

## Dashboard description:
The dashboard shown above has two plots -each showing the number of Free, Paid Games in the Steam Game Platform and the Maximum number of Achievement in games with respect to the month of release of games.

The above chart is interactive. If the mouse pointer is clicked and dragged to select an interval of months in the left chart, it can be seen that the line chart in the right changes according to the interval selected.

The left chart shows the number of free and paid games in the Steam Game platform shown in light blue (Paid games) and dark blue(free games) bars. The number of free and paid games in each month of release of game can also be observed.

The right chart shows the maximum achievement count of games in the Steam Game platform as a line chart. The peaks represent the maximum achievement count in each release month of the games.

For instance, the left chart shows that during the release months- June and August, the number of free games were high. If the selection is made for months- June, July,August in the left chart, the corresponding maximum achievement count for those months can be seen from the line chart shown on the right. The right chart shows that the achievement count peaked during the month of June with 1150 within the selection interval from June to August.

Similar analysis can be made for the maximum achievement count for paid games, maximum achievement count for free games and comparing them with their month of release can be done using the dashboard and the trends according to the release month, achievement count, free/paid games can be visualized. This might be useful for the Steam Game Platform to improve their gaming service.

## Contextual Dataset:
### Steam Video Game Hours Played Dataset
DATASET: An interesting contextual dataset that would add context to the Steam Game Dataset used would be ' Steam Video Game Hours Played' dataset.

CITATION: Steam Video Games. (2017, March 9). Kaggle. https://www.kaggle.com/datasets/tamber/steam-video-games

LICENSE: http://opendatacommons.org/licenses/dbcl/1.0/

URL: https://www.kaggle.com/datasets/tamber/steam-video-games?resource=download


### Description-Contextual Dataset:
The Steam Video Game Hours Played dataset includes a list of the user behaviors with columns- user ID, game title, the behavior including 'purchase' and 'play'.

The dataset used for the project was the Steam Game dataset which includes information about the games in the Steam Game platform with hardware requirements,  different genre of games, currency used, recommendation count, release date, response count,achievement count etc., but there is no information about the behavior of users in the platform with respect to the number of hours played by users.

The contextual dataset- 'Steam Video Game Hours Played' which includes information about the number of hours played by gamers in the Steam Game platform would be useful to visualize the overall user activity and the specification of the Steam Game platform along with the original dataset, Steam Game Dataset.
This would benefit the developers of the Steam Game platform to improve their overall interface and gamers would be able to see game statistics and individual statistics.

### Team Members:
Vasantha Chandrasekaran and Saahil Hiranandani